SIPS/sip-contract-storage.md

```diff
@@ -0,0 +1,118 @@
+>
+>     2. A potentially better solution (though it could get ugly) is having an available contract mapping. Initially the mapping is empty but it could be added to by a contract that allows other named contracts in the `AddressResolver` to set/get on its behalf. So for the above example, `Issuer` would have something added to itself that says `Storage.addContractMapping("Issuer", "Burner")` that would allow `Burner` to pass through `Issuer` as a key and still have write access to that space. This isn't great because then `Burner` needs to keep a reference to the `"Issuer"` storage key, but is manageable. Any other suggestions?
+
+# Rationale
+
+<!--This is where you explain the reasoning behind how you propose to solve the problem. Why did you propose to implement the change in this way, what were the considerations and trade-offs. The rationale fleshes out what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+
+The current `EternalStorage` contract is quite flexible in how it stores state, it makes sense to double down on this useful pattern, hence a new version based off of it (but not extending it as we don't want to use the `State` mixin here). One trade-off of this approach however is the storage of packed structs. If a struct condenses it's entries down to maximize space: e.g. [`SystemStatus.Suspension`](https://github.com/Synthetixio/synthetix/blob/v2.21.15/contracts/SystemStatus.sol#L17-L22) (which uses a `bool` that is `uint8` in Solidity, combined with a `uint248` reason code which makes up the remainder of the 32 bytes in the storage slot), then it would be less efficient to store the individual components into two separate storage slots.
+
+Unfortunately, there's no easy solution to generalizing the storage of structs. The usage of this storage contract will come down to whether or not the contract in question can efficiently compress and decompress its required data before going in and out of the storage contract.
+
+Additionally, there will be a slightly higher gas cost when persisting storage now as each contract will need to do a cross-contract call both a) to the new `Storage` contract and then b) from the new `Storage` contract to the `AddressResolver` to ascertain if `msg.sender` is indeed the address it expects from the `AddressResolver`. This second step cannot be alleviated by having `Storage` use `MixinResolver` as other Synthetix contracts do because `Storage` is not an upgradable contract, and thus we can't hard-code the names of the contracts it needs to store in its cache.
```

Comment:
new `Storage` contract could persist the list of `mapping(Address -> bytes32) addressToContractNames` internally from AddressResolver to optimise the cost there. The contract names it needs to store in its cache could be appended by `onlyOwner` action vs defining them at contract deployment. 

Unfortunately it seems that by moving to the eternalStorage pattern, you would lose the ability to compact structs. Looking at some solutions to this, 
https://ethereum.stackexchange.com/questions/64127/why-is-the-eternal-storage-approach-still-used-if-it-hugely-increases-gas-usage

Reply:
It may be possible to retain the gas optimisation and compact a struct's variables into a 256 bits slot and store the encoded result in `mapping(bytes32 => bytes32) internal Bytes32Storage;` using assembly to encode the variables before storing in the eternalStorage and decoding to reconstruct the struct. 

For example, the gas cost of storing the struct and the encoded bytes32 are the same here as the uint variables are compacted. 

```solidity
    struct Entry {
        uint48 a;
        uint8 b;
        uint8 c;
    }

    function _encode(uint48 a, uint8 b, uint8 c) internal pure returns (bytes32 x) {
        assembly {
        let y := 0
            mstore(0x20, c)
            mstore(0x18, b)
            mstore(0x10, a)
            x := mload(0x20)
        }
    }

    function decode(bytes32 x) public pure returns (uint48 a, uint8 b, uint8 c) {
        assembly {
            c := x
            mstore(0x10, x)
            a := mload(0)
            mstore(0x8, x)
            b := mload(0)
        } 
    }
```

- Another method to store all of a struct's variables is to use `abi.encode` and `abi.decode` to decode the expected variables of the struct, and storing the result bytes on the eternal storage. It should save more gas than making multiple cross contract and store operations to save each individual variable separately. 

However testing this showed higher gas usage when storing the encoded bytes compared to storing the struct internally in a mapping.